esp/services/ws_dfu/ws_dfuXRefService.hpp

```diff
@@ -26,116 +26,144 @@
-       xRefNode->Link();
-       m_pNodeQueue.enqueue(xRefNode);
-       m_sem.signal();
-   }
-
-   ~CXRefExBuilderThread(){DBGLOG("Destroyed XRef thread");}
+    IMPLEMENT_IINTERFACE;
+
+    CXRefExBuilderThread() { };
+    ~CXRefExBuilderThread(){DBGLOG("Destroyed XRef thread");};
+
+    virtual void queueRequest(IXRefNode* xRefNode, const char* cluster)
+    {
+        if (!xRefNode || isEmptyString(cluster))
+            return;
+
+        xRefNode->setCluster(cluster);
+        writeNodeQueue(xRefNode);
+        m_sem.signal();
+    }
+
     virtual int run()
     {
         Link();
-        while(m_bRun)
+        while (!stopThread)
         {
             m_sem.wait();
-            if (m_pNodeQueue.ordinality() != 0)
-                RunXRef();
+            runXRef();
         }
         Release();
         return 0;
     }
 
-    void RunXRef()
+    void runXRef()
     {
-
         //catch all exceptions so we can signal for the new build to start
-        try{
-            SetRunningStatus(true);
-            Owned<IXRefNode> xRefNode  = (IXRefNode*)m_pNodeQueue.dequeue();
-            _CurrentClusterName.clear();
-            xRefNode->getCluster(_CurrentClusterName);
-            if (xRefNode->useSasha()) // if sasha processing just set submitted
-                xRefNode->setStatus("Submitted");
-            else {
-                Owned<IPropertyTree> tree = runXRefCluster(_CurrentClusterName.str(), xRefNode);
-                DBGLOG("finished run");
+        try
+        {
+            while (true)
+            {
+                Owned<IXRefNode> xRefNode  = readNodeQueue();
+                if (!xRefNode)
+                    break;
+
+                if (xRefNode->useSasha()) // if sasha processing just set submitted
+                    xRefNode->setStatus("Submitted");
+                else
+                {
+                    setRunningStatus(true);
+                    Owned<IPropertyTree> tree = runXRefCluster(currentClusterName.str(), xRefNode);
+                    DBGLOG("finished run XRef for %s", currentClusterName.str());
+                    clearCurrentClusterName();
+                    setRunningStatus(false);
+                }
             }
-            SetRunningStatus(false);
         }
         catch(IException* e)
         {
             StringBuffer errorStr;
             e->errorMessage(errorStr);
-            ERRLOG("Exception thrown while running XREF: %s",errorStr.str());
+            ERRLOG("Exception thrown while running XREF: %s", errorStr.str());
             e->Release();
         }
         catch(...)
         {
             ERRLOG("Unknown Exception thrown from XREF");
         }
-        //Signal that we are ready to process another job if there is one....
-        m_sem.signal();
-
     }
 
-    virtual bool IsRunning()
+    virtual bool isRunning()
     {
-        CriticalSection(_boolMutex);
-        return bRunning;
+        CriticalBlock b(critRunningStatus);
```

Comment:
Bit pointless using a CriticalBlock here, I think. If there's a race condition where someone can be changing this value around the same time as you are calling it, you are still going to get a random value.